docs/proposals/failover/optimize-cluster-failover.md

```diff
@@ -0,0 +1,635 @@
+3. To facilitate the distinction of concepts, we can collectively refer to the migrations triggered by taint{effect: NoExecute} as migrations or passive migrations, and the migrations triggered by taint{effect: PreferNoExecute} as active migrations, because taint{effect: PreferNoExecute} allows applications to stay in place without any toleration.
+
+For the API definition of `PP/CPP.spec.failover.cluster`, please refer to the next section: Application-level cluster failover strategy design.
+
+### Application-level cluster failover strategy design
+
+We plan to introduce an application-level cluster failover strategy to configure an active migration policy for the PreferNoExecute taint.
+
+Introducing new fields in PropagationPolicy/ClusterPropagationPolicy `.spec.failover` API as follows:
+
+```go
+// FailoverBehavior indicates failover behaviors in case of an application or
+// cluster failure.
+type FailoverBehavior struct {
+    …
+
+    // Cluster indicates failover behaviors in case of cluster failure.
+    // The cluster failure here refers to the cluster taint that contains
+    // the taint whose effect is PreferNoExecute.
+    //
+    // PreferNoExecute taints do not need to consider the setting of tolerations.
+    //
+    // If this value is nil, failover is disabled.
+    // +optional
+    Cluster *ClusterFailoverBehavior `json:"cluster,omitempty"`
+}
+
+// ClusterFailoverBehavior indicates cluster failover behaviors.
+type ClusterFailoverBehavior struct {
+    // PurgeMode represents how to deal with the legacy applications on the
+    // cluster from which the application is migrated.
+    // Valid options are "Directly", "Gracefully".
+    // Defaults to "Gracefully".
+    // +kubebuilder:validation:Enum=Directly;Gracefully
+    // +kubebuilder:default=Gracefully
+    // +optional
+    PurgeMode PurgeMode `json:"purgeMode,omitempty"`
+
+    // TolerationSeconds represents the time system should wait for
+    // performing failover process after the PreferNoExecute
+    // taint occurs on the cluster.
+    // If not specified, system will immediately perform failover process.
+    // Defaults to 300s.
+    // +kubebuilder:default=300
+    // +optional
+    TolerationSeconds *int32 `json:"tolerationSeconds,omitempty"`
```

Comment:
The `TolerationSeconds` in PropagationPolicy and `AddOnMatchSeconds` in `ClusterTaintPolicy` both delay the migration. Maybe keep one would be enough?